ibmcloud/README.md

```diff
@@ -0,0 +1,269 @@
+
+If ansible fails for some reason, you can rerun the ansible playbook as follows.
+```
+cd ansible
+ansible-playbook -i ./inventory -u root ./playbook.yml
+```
+
+When ansible fails, Terraform does not execute the setup script for Kubernetes. In this case, you can manually run it as follows. Note that you do not need to run this script manually, when everything goes well.
+
+```
+./scripts/setup.sh --bastion <floating IP of the worker node> --control-plane <IP address of the control-plane node> --workers  <IP address of the worker node>
+```
+
+When two VSIs are successfully provisioned, a floating IP address is assigned to the worker node. Please use the floating IP address to access the worker node from the Internet.
+
+## Build a pod VM image
+
+You need to build a pod VM image for peer pod VMs. A pod VM image contains the following components.
+
+* Kata agent
+* Agent protocol forwarder
+* skopeo
+* umoci
+
+The build scripts are located in [ibmcloud/image](./image). The prerequisite software to build a pod VM image is already installed in the worker node by [the Ansible playbook](terraform/cluster/ansible/playbook.yml) for convenience. Note that building a pod VM image on a worker node is not recommended for production, and we need to build a pod VM image somewhere secure to protect workloads running in a peer pod VM.
+
+The following command builds a custom VM image. A new QCOW2 file with prefix `podvm-` will be created in the current directory.
+
+```
+cd ibmcloud/image
+make build
+```
+
+You need to configure Cloud Object Storage (COS) to upload your custom VM image.
+
+https://cloud.ibm.com/objectstorage/
+
+First, create a COS service instance if you have not create one. Then, create a COS bucket with the COS instance. The COS service instance and bucket names are necessary to upload a custom VM image.
+
+The following environment variables are necessary to be set before executing the image upload script. You can change IBMCLOUD_COS_REGION if you prefer another region. In this case, you also want to change IBMCLOUD_COS_SERVICE_ENDPOINT to one of endpoints listed at [https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-endpoints](https://cloud.ibm.com/docs/cloud-object-storage?topic=cloud-object-storage-endpoints).
+
+```
+export IBMCLOUD_API_KEY=<your API key>
+export IBMCLOUD_COS_SERVICE_INSTANCE=<COS service instance name>
+export IBMCLOUD_COS_BUCKET=<COS bucket name>
+export IBMCLOUD_COS_REGION=jp-tok
+export IBMCLOUD_COS_SERVICE_ENDPOINT=https://s3.jp-tok.cloud-object-storage.appdomain.cloud
+```
+
+Next, you need to grant access to COS to import images as described at [https://cloud.ibm.com/docs/vpc?topic=vpc-object-storage-prereq&interface=cli](https://cloud.ibm.com/docs/vpc?topic=vpc-object-storage-prereq&interface=cli).
+
+```
+ibmcloud login -r jp-tok -apikey $api_key
```

Comment:
Should this be `$IBMCLOUD_API_KEY` rather than `$api_key` as we've just set it?

Reply:
I've just spotted that `ibmcloud/image/login.sh` is expecting to use `$api_key`:
```sh
if [[ -n "$api_key" ]]; then
        opts=(--apikey "$api_key")
```
so maybe there should be instructions to set this instead of/as well as `IBMCLOUD_API_KEY`?

Reply:
The description in README is wrong. It should be `$IBMCLOUD_API_KEY` as you pointed out. I'll fix it.

In `login.sh`, `$api_key` is defined as `$api_key=${IBMCLOUD_API_KEY-}`, so `$api_key` is a local variable in the script.